### ISSUE TITLE
Provide a way to set permissions on authorized_keys within a remote session

### ISSUE THREAD

**Issue by lucastheisen** [PowerShell/Win32-OpenSSH#747]
Please answer the following

**If it is a terminal issue then please go through wiki
https://github.com/PowerShell/Win32-OpenSSH/wiki/TTY-PTY-support-in-Windows-OpenSSH**

No

**"OpenSSH for Windows" version**
```
 PS C:\Users\foobar> ((Get-Item (Get-Command sshd).Source).VersionInfo.FileVersion)
0.0.14.0
```
**Server OperatingSystem**
```
PS C:\Users\foobar> ((Get-ItemProperty "HKLM:\SOFTWARE\Microsoft\Windows nt\CurrentVersion\" -Name ProductName).ProductName)                                                                               
Windows 10 Pro
```
**Client OperatingSystem**
```
PS C:\WINDOWS\system32> ((Get-ItemProperty "HKLM:\SOFTWARE\Microsoft\Windows nt\CurrentVersion\" -Name ProductName).ProductName)
Windows 10 Pro
```
**What is failing**

Set-SSHKeyPermissions.ps1.  More specifically, a typical usage scenario would be ssh into a new server using your password, create the .ssh/authorized_key file, then add your pub key to that file.  Then set permissions as needed, log out, log back in and verify the key worked.  But since Set-SSHKeyPermissions.ps1 fails, this cannot be done.  It appears there are 2 real problems here, the first is what I see in my actual output: `The SSHD service must exist in order to grant permissions to the virtual user "NT Service\SSHD"`, the second would be that even after that, it would fail with `The process does not possess the 'SeSecurityPrivilege' privilege which is required for this operation.` since the shell is not run as admin.  Perhaps I am doing something wrong?

**Expected output**
```
PS C:\WINDOWS\system32> Set-SSHKeyPermissions.ps1

*****************************************
https://CloudyWindows.com SSH Utility Script - Set-SSHKeyPermissions.ps1
Run 'Set-SSHKeyPermissions' at a PowerShell Prompt each time you add an authorized_keys file
or set it up in the task scheduler.
Searching for and granting read permissions to SSHD Service for all user authorized_keys files...
    Giving read permissions to "NT Service\SSHD" for "C:\Users\foobar\.ssh\authorized_keys"
*****************************************
```
**Actual output**
```
PS C:\Users\foobar> Set-SSHKeyPermissions.ps1

*****************************************
https://CloudyWindows.com SSH Utility Script - Set-SSHKeyPermissions.ps1
Run 'Set-SSHKeyPermissions' at a PowerShell Prompt each time you add an authorized_keys file
or set it up in the task scheduler.
Searching for and granting read permissions to SSHD Service for all user authorized_keys files...
The SSHD service must exist in order to grant permissions to the virtual user "NT Service\SSHD"
At C:\Program Files\OpenSSH-Win64\Set-SSHKeyPermissions.ps1:12 char:3
+   Throw "The SSHD service must exist in order to grant permissions to ...
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (The SSHD servic...T Service\SSHD":String) [], RuntimeException
    + FullyQualifiedErrorId : The SSHD service must exist in order to grant permissions to the virtual user "NT Servic
```

**Comment by bingbing8** [comment]
@lucastheisen Is your OpenSSH binaries v0.0.14.0 or v0.0.15.0?
Is Set-SSHKeyPermissions.ps1 your own script?
To set permission of authorized_keys, you need to call Fix-AuthorizedKeyPermissions. Can you try and let me know?

**Comment by lucastheisen** [comment]
@bingbing8 , I am using v0.0.14.0.  `Set-SSHKeyPermissions.ps1` is a script that comes as part of the [chocolatey distro](https://chocolatey.org/packages/openssh) you link to in [your documentation](https://github.com/PowerShell/Win32-OpenSSH/wiki/Win32-OpenSSH-Automated-Install-and-Upgrade-using-Chocolatey).   Also, `Fix-AuthorizedKeyPermissions` does not exist in my installation.  Since you just released v0.0.15.0, maybe @DarwinJS will update the chocolatey distro soon...

**Comment by manojampalam** [comment]
@lucastheisen we'll have a solution for this in next release.

For now, you'll need to use icacls.exe to set the appropriate permissions as detailed here - https://github.com/PowerShell/Win32-OpenSSH/wiki/Security-protection-of-various-files-in-Win32-OpenSSH 
See section on authorized_keys

**Comment by DarwinJS** [comment]
Set-SSHKeyPermissions.ps1 is part of the chocolatey package and addressed the problem of setting all user based keys during install and was designed to be run by an admin user in order to set permissions on all user keys.

I am working on version 0.0.15.0 of the chocolatey script and Set-SSHKeyPermissions.ps1 will be retired in favor of the utility script FixHostFilePermissions.ps1 that ships with the zip.

**Comment by bingbing8** [comment]
@lucastheisen can you try the [utility scripts](https://github.com/PowerShell/Win32-OpenSSH/wiki/OpenSSH-utility-scripts-to-fix-file-permissions) in [v0.0.16.0](https://github.com/PowerShell/Win32-OpenSSH/releases/tag/v0.0.16.0) to see if it can meet your requirement to set permissions on authorized_keys?
```
#   fix permissions on sshd_config, authorized_keys, and host keys
.\FixHostFilePermissions.ps1
```
or
```
Import-Module .\OpenSSHUtils.psd1 -Force
# fix permissions on a specified authorized_key
Repair-AuthorizedKeyPermission -FilePath C:\Users\sshtest_ssouser\.ssh\authorized_keys
```

**Comment by lucastheisen** [comment]
@bingbing8 , the first suggestion ends up spitting out a bunch of errors because it attempts to update permissions on all the files (including the sshd host keys which the non-admin user does not have permission to do).  It will, however, eventually run the permission update on the key in question...  But all the errors are rather distracting.  Perhaps that script could be updated to check if being run as a non-admin user, and if so, only attempt to set the permissions on that users key.

Your second approach should probably read:
```
Import-Module "C:\Program Files\OpenSSH-Win64\OpenSSHUtils.psd1" -Force
```
Then you can run the:
```
Repair-AuthorizedKeyPermission -FilePath C:\Users\foobar\.ssh\authorized_keys
```
and it works!

While this does make it _possible_ to accomplish, it is still rather difficult.  The following is my incantation:
```
#!/bin/bash

IDENTITY=~/.ssh/id_rsa.pub
while getopts "i:" flag; do
    case $flag in
        i ) IDENTITY="$OPTARG" ;;
    esac
done
shift $((OPTIND-1))
DESTINATION=$1
if [ -z $DESTINATION ]; then
    echo "destination required"
    exit 1
fi

read -r -d '' POWERSHELL_SCRIPT << 'EOF'
[Console]::In.ReadLine() | Out-File .ssh\authorized_keys -Encoding ASCII
Import-Module 'C:\Program Files\OpenSSH-Win64\OpenSSHUtils.psd1' -Force
Repair-AuthorizedKeyPermission -FilePath .ssh\authorized_keys -Confirm:$false
EOF
ENCODED_COMMAND=$(echo -n "$POWERSHELL_SCRIPT" | iconv -f UTF-8 -t UTF-16LE | base64 -w 0)
echo "$ENCODED_COMMAND"

REMOTE_COMMAND="powershell -EncodedCommand $ENCODED_COMMAND"
echo "$REMOTE_COMMAND"

cat "$IDENTITY" | ssh "$DESTINATION" "$REMOTE_COMMAND"
```

This would not be required if `ssh-copy-id` worked, but:
```
# ssh-copy-id -i .ssh/id_rsa.pub foobar@ljt530
/usr/bin/ssh-copy-id: INFO: Source of key(s) to be installed: ".ssh/id_rsa.pub"
/usr/bin/ssh-copy-id: INFO: attempting to log in with the new key(s), to filter out any that are already installed
/usr/bin/ssh-copy-id: INFO: 1 key(s) remain to be installed -- if you are prompted now it is to install the new keys
foobar@ljt530's password:
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQCsQLO8xetmJmbTthVW7DT9ecqMw/yBQTm1sMA/Au11CLubU/pzJjkGUWUfMHyqWHt1YI1wtna2z3033T+o8QLHToclS6vnh47VSl7tVxtiz6NmwZcOD22J48rR480WDKJyk6Ck7LtFMu+T2TBidL+Q1HXJZewi2xmKC5FlWPF2nAmd+xTfJ5FtRUqf03FbQndVpbeTMkPWDB4d+qSSK57Xl8vwaWod0UhxFRpxTSGDs/mMbbbAzbqDOIIRg4ab6kb22+fecMVV1eRJQVkQdRlmNcsIOQt3Zhjk3E9yJVNtBtQ9j4hQGBLiuUwpoNXnrlWnrQpeHMMhiOIiAriFYqPjXEn+9f/fdHKnqtHwZunbu1Tae3D6DmEY4gHg3N0/jo7B1YmW/ADmaEhFRx35TWN6LrYO2eGOJI4TEi+aGsynVoZtavsH6nKHg/eLso3hmX0MwOTvH9d/TkCtgwMRxe12GNlnDWbZ2gtOmNQKj9MIJGaxmN/lJNoLhPjftqYOiSb0yuKWzAoLflDgOMzmnKlfABk+C8uRfA+kGuWlPJ+TBpdxCMgEZBVuPp5GL3BSI8XOETdRQqTv0dfNcIZZDQ23UzeiKwvTxlRFUOKlfrHyyqG7IhLZkj2zxcCtnuQbJiyKLAKuJ9t4u2K4jJVChuw/43V0doMmWO+2Rb4U5RicLQ== ltheisen@ltserver
'exec' is not recognized as an internal or external command,
operable program or batch file.
The system cannot find the path specified.
```

In any case, thank you for fixing this.  Great work!
